include/quic/stream.hpp

```diff
@@ -95,22 +94,159 @@ namespace oxen::quic
+        }
+
+      private:
+        // Implementations classes for send_chunks()
+
+        // chunk_sender: When sending chunks we construct *one* of these, then share its ownership
+        // across all the chunks in flight.  When each individual chunk gets destroyed, it called
+        // back into this to queue the next chunk, which this class then sends into the stream.
+        //
+        // Container -- can be a value or a pointer, but not a reference.
+        template <typename Container>
+        struct chunk_sender : std::enable_shared_from_this<chunk_sender<Container>>
+        {
+            static_assert(!std::is_reference_v<Container>, "chunk_sender requires a value or pointer, not a reference");
+
+            static constexpr bool is_pointer = std::is_pointer_v<Container> ||
+                                               is_instantiation<std::unique_ptr, Container> ||
+                                               is_instantiation<std::shared_ptr, Container>;
+
+            using chunk_callback_t = std::function<Container(const Stream&)>;
+            using done_callback_t = std::function<void(Stream&)>;
+
+            template <typename... Args>
+            static void make(int initial_queue, Args&&... args)
+            {
+                std::shared_ptr<chunk_sender<Container>> cs{new chunk_sender<Container>(std::forward<Args>(args)...)};
+                for (int i = 0; i < initial_queue; i++)
+                    cs->queue_next_chunk();
+            }
+
+          private:
+            // This is instantiated for each chunk, contains the chunk data itself, and is what we
+            // put into the keep_alive; during destruction, we queue the next chunk.
+            struct single_chunk
+            {
+              private:
+                std::shared_ptr<chunk_sender> _chunks;
+                Container _data;
+
+              public:
+                single_chunk(chunk_sender& cs, Container&& d) : _chunks{cs.shared_from_this()}, _data{std::move(d)} {}
+                ~single_chunk() { _chunks->queue_next_chunk(); }
+
+                bstring_view view() const
+                {
+                    if constexpr (is_pointer)
+                    {
+                        static_assert(sizeof(*_data->data()) == 1, "chunk_sender requires bytes data");
+                        return {reinterpret_cast<const std::byte*>(_data->data()), _data->size()};
+                    }
+                    else
+                    {
+                        static_assert(sizeof(*_data.data()) == 1, "chunk_sender requires bytes data");
+                        return {reinterpret_cast<const std::byte*>(_data.data()), _data.size()};
+                    }
+                }
+            };
+
+            chunk_sender(Stream& s, chunk_callback_t next, done_callback_t done) :
+                    str{s}, next_chunk{std::move(next)}, done{std::move(done)}
+            {
+                assert(next_chunk);
```

Comment:
static_assert?

Reply:
Not compile-time checkable.

If this is false here, the caller did something wrong enough that it isn't something we want to be catchable.